packages/data-wizard/__tests__/dataset/data-frame.test.ts

```diff
@@ -880,3 +880,186 @@ describe('DataFrame data with fillValue', () => {
+  test('1D: basic type', () => {
+    const df = new DataFrame(201, { columnTypes: ['string'] });
+    expect(df.data).toStrictEqual([['201']]);
+    expect(df.colData).toStrictEqual([['201']]);
+  });
+
+  test('1D: basic type with extra indexes and columns', () => {
+    const df = new DataFrame(201, { indexes: ['kk', 'bb'], columns: [5, 6], columnTypes: ['string', 'integer'] });
+    expect(df.data).toStrictEqual([
+      ['201', 201],
+      ['201', 201],
+    ]);
+    expect(df.colData).toStrictEqual([
+      ['201', 201],
+      ['201', 201],
+    ]);
+  });
+
+  test('1D: array', () => {
+    const df = new DataFrame([201, '201', 3], { columnTypes: ['string'] });
+    expect(df.data).toStrictEqual([['201'], ['201'], ['3']]);
+    expect(df.colData).toStrictEqual([['201', '201', '3']]);
+  });
+
+  test('1D: array with extra indexes and columns', () => {
+    const df = new DataFrame([1, 2, 201], { indexes: ['b', 'c', 'a'], columns: ['col'], columnTypes: ['string'] });
+    expect(df.data).toStrictEqual([['1'], ['2'], ['201']]);
+    expect(df.colData).toStrictEqual([['1', '2', '201']]);
+  });
+
+  test('1D: object', () => {
+    const df = new DataFrame({ a: 1, b: '201', c: 3 }, { columnTypes: ['string', 'string', 'string'] });
+    expect(df.data).toStrictEqual([['1', '201', '3']]);
+    expect(df.colData).toStrictEqual([['1'], ['201'], ['3']]);
+  });
+
+  test('1D: object with extra indexes and columns', () => {
+    const df = new DataFrame(
+      { a: 1, b: 2, c: 201 },
+      { indexes: ['idx1'], columns: ['c', 'b'], columnTypes: ['', 'string'] }
+    );
+    expect(df.data).toStrictEqual([[201, '2']]);
+    expect(df.colData).toStrictEqual([[201], ['2']]);
+  });
+
+  test('2D: array', () => {
+    const df = new DataFrame(
+      [
+        [1, 4],
+        [2, '201'],
+        [3, 6],
+      ],
+      { columnTypes: ['string', 'integer'] }
+    );
+    expect(df.data).toStrictEqual([
+      ['1', 4],
+      ['2', 201],
+      ['3', 6],
+    ]);
+    expect(df.colData).toStrictEqual([
+      ['1', '2', '3'],
+      [4, 201, 6],
+    ]);
+  });
+
+  test('2D: array with extra indexes and columns', () => {
+    const df = new DataFrame(
+      [
+        [1, 4],
+        [2, 5],
+        [3, 6],
+      ],
+      {
+        indexes: ['a', 'b', 'c'],
+        columns: ['col1', 'col2'],
+        columnTypes: ['string', 'integer'],
+      }
+    );
+    expect(df.axes).toStrictEqual([
+      ['a', 'b', 'c'],
+      ['col1', 'col2'],
+    ]);
+    expect(df.data).toStrictEqual([
+      ['1', 4],
+      ['2', 5],
+      ['3', 6],
+    ]);
+    expect(df.colData).toStrictEqual([
+      ['1', '2', '3'],
+      [4, 5, 6],
+    ]);
+  });
+
+  test('2D: object array', () => {
+    const df = new DataFrame(
+      [
+        { a: '1', b: '2', c: '3.1', d: 1 },
+        { a: '1', b: '2', d: '4' },
+        { c: '1', d: '2019-02-03', e: 0, f: 201 },
+      ],
+      {
+        columnTypes: ['null', 'boolean', 'float', 'date', 'string'],
```

Comment:
i think it's difficult to understand the correspondence between  indexes in `columnTypes`  and fields

Reply:
`columnTypes` and `columns` are one–to–one correspondence. If you understand `columns`, you can understand `columnTypes`. It's hard to specify data types by `{a: 'string', b: 'boolean'}`, because data such as a two-dimensional array has no column fields.